MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,322 @@
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//      /\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Although we have not removed node 1 and node 2 from the MMR, we cannot yet remove node 3 as we require node 3 for
+// the proof of node 6. Let's prune 4 and 5.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//         /\  /\ \    /\  /\  /\
+//        /\/\/\/\/\  /\/\/\/\/\/\
+//
+// Now we removed 3 from the MMR
+
+use crate::merklenode::{Hashable, MerkleNode, ObjectHash};
+use std::collections::HashMap;
+
+pub struct MerkleMountainRange<T>
+where T: Hashable
+{
+    // todo convert these to a bitmap
+    mmr: Vec<MerkleNode>,
+    data: HashMap<ObjectHash, T>,
+}
+
+impl<T> MerkleMountainRange<T>
+where T: Hashable
+{
+    /// This function creates a new empty Merkle Mountain Range
+    pub fn new() -> MerkleMountainRange<T> {
+        MerkleMountainRange { mmr: Vec::new(), data: HashMap::new() }
+    }
+
+    /// This function returns a reference to the data stored in the mmr
+    /// It will return none if the hash does not exist
+    pub fn get_object(&self, hash: ObjectHash) -> Option<&T> {
+        self.data.get(&hash)
+    }
+
+    /// This function returns a mut reference to the data stored in the mmr
+    /// It will return none if the hash does not exist
+    pub fn get_mut_object(&mut self, hash: ObjectHash) -> Option<&mut T> {
+        self.data.get_mut(&hash)
+    }
+
+    /// This function returns the hash proof tree of a given hash.
+    /// If the given hash is not in the tree, the vec will be empty.
+    /// The Vec will be created in form of the child-child-parent
+    pub fn get_hash_proof(&self, hash: ObjectHash) -> Vec<ObjectHash> {
+        let mut result = Vec::new();
+        let mut i = self.mmr.len();
+        for counter in 0..self.mmr.len() {
+            if self.mmr[counter].hash == hash {
+                result.push(self.mmr[i].hash.clone());
+                i = counter;
+                break;
+            }
+        }
+        i = peer_index(i as u64) as usize;
+        while i < self.mmr.len() {
+            result.push(self.mmr[i].hash.clone());
+            i += 1;
+            result.push(self.mmr[i].hash.clone());
+
+            i = peer_index(i as u64) as usize;
+        }
+        result
+    }
+
+    /// This function returns the peak height of the mmr
+    pub fn get_peak(&self) -> u32 {
```

Comment:
```suggestion
    pub fn get_peak_height(&self) -> u32 {
```

also I suggest using `usize` everywhere.